# Data gathering

## 1) Artist information

In [1]:
import pandas as pd
import numpy as np
#!pip install pygeocoder
from pygeocoder import Geocoder #If you want to follow the geocoding later, you will need your own Google Maps API key
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
artists= pd.read_csv('Musicbrainz/Tables_used/artist.txt',sep='\t', header=None, engine='c', usecols=[0,1,2,11,17])
artists.columns = ['artist_id','artist_mbid','artist_name','start_area1', 'start_area2']
artists.head()

artist_id                           artist_mbid      artist_name  \
0     805192  8972b1c1-6482-4750-b51f-596d2edea8b1            WIK▲N   
1     371203  49add228-eac5-4de8-836c-d75cde7369c3      Pete Moutso   
2     273232  c112a400-af49-4665-8bba-741531d962a1          Zachary   
3     101060  ca3f3ee1-c4a7-4bac-a16a-0b888a396c6b  The Silhouettes   
4     145773  7b4a548e-a01a-49b7-82e7-b49efeb9732c     Aric Leavitt   

   start_area1  start_area2  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3        222.0       7707.0  
4          NaN          NaN

In [3]:
#Let's see how many artists we have:
artists['artist_id'].nunique()

1476425

In [4]:
#How much info we have for each artist?
artists.isnull().sum(axis=0)

artist_id            0
artist_mbid          0
artist_name          8
start_area1     808442
start_area2    1274001
dtype: int64

What are the "start_area1" and "start_area2"? If we look at Musicbrainz's field description for each artist (https://musicbrainz.org/doc/Artist), we can see that:

Area: The artist area, as the name suggests, indicates the area with which an artist is primarily identified with. It is often, but not always, its birth/formation country.

We will keep this information as the artist's origin for later.

We need to incorporate as well the table called "artist credit", which gives us the artist credit_id. We will use this field to join later on each release with its artist:

In [5]:
artists_credit= pd.read_csv('Musicbrainz/Tables_used/artist_credit_name.txt',sep='\t', header=None, engine='c', usecols=[0,2,3])
artists_credit.columns = ['credit_id','artist_id','artist_name']
artists_credit.head()

credit_id  artist_id             artist_name
0     578352     578352           Gustav Ruppke
1     273232     273232                 Zachary
2     153193     153193  The High Level Ranters
3      32262      32262        Georges Brassens
4    1389968    1171184    Harvard of the South

In [6]:
#Let's join the artists with their credit id and verify that the matching is good:
df = pd.merge(artists, artists_credit, how='left', on='artist_id')
df.head()

artist_id                           artist_mbid    artist_name_x  \
0     805192  8972b1c1-6482-4750-b51f-596d2edea8b1            WIK▲N   
1     371203  49add228-eac5-4de8-836c-d75cde7369c3      Pete Moutso   
2     273232  c112a400-af49-4665-8bba-741531d962a1          Zachary   
3     101060  ca3f3ee1-c4a7-4bac-a16a-0b888a396c6b  The Silhouettes   
4     145773  7b4a548e-a01a-49b7-82e7-b49efeb9732c     Aric Leavitt   

   start_area1  start_area2  credit_id    artist_name_y  
0          NaN          NaN   822846.0            WIK▲N  
1          NaN          NaN        NaN              NaN  
2          NaN          NaN   273232.0          Zachary  
3        222.0       7707.0   101060.0  The Silhouettes  
4          NaN          NaN   145773.0     Aric Leavitt

In [7]:
#It looks like it makes sense. Please note that the credit id is sometimes equal to the artist_id, but not always:
df['check'] = df['artist_id'] - df['credit_id']
df['check'].nunique()

1270628

In [8]:
df.isnull().sum(axis=0)

artist_id              0
artist_mbid            0
artist_name_x         15
start_area1      1120376
start_area2      2109027
credit_id         461241
artist_name_y     461253
check             461241
dtype: int64

In [9]:
#We can now get rid of check and the duplicate artist_name column:
df.drop(labels=['check','artist_name_y'], axis=1, inplace=True)
df.head()

artist_id                           artist_mbid    artist_name_x  \
0     805192  8972b1c1-6482-4750-b51f-596d2edea8b1            WIK▲N   
1     371203  49add228-eac5-4de8-836c-d75cde7369c3      Pete Moutso   
2     273232  c112a400-af49-4665-8bba-741531d962a1          Zachary   
3     101060  ca3f3ee1-c4a7-4bac-a16a-0b888a396c6b  The Silhouettes   
4     145773  7b4a548e-a01a-49b7-82e7-b49efeb9732c     Aric Leavitt   

   start_area1  start_area2  credit_id  
0          NaN          NaN   822846.0  
1          NaN          NaN        NaN  
2          NaN          NaN   273232.0  
3        222.0       7707.0   101060.0  
4          NaN          NaN   145773.0

## 2) Release information

The objective of this project is to visualize when each artist released for the first time a certain CD/Album/Single etc.

If we look at the "releases" table:

In [10]:
releases = pd.read_csv('Musicbrainz/Tables_used/release.txt',sep='\t', header=None, engine='c', usecols=[0,2,3])
releases.columns = ['release_id','release_group','credit_id']
releases.head()

release_id           release_group  credit_id
0           9       A Sorta Fairytale         60
1          10       A Sorta Fairytale         60
2          11       Glory of the 80's         60
3          12  Silent All These Years         60
4          26                  Demons      20211

We can see, in the first 2 rows, that the same CD/Album can be released/remastered many times. According to Musicbrainz's field description for each release (https://musicbrainz.org/doc/Release):

"A MusicBrainz release represents the unique release (i.e. issuing) of a product on a specific date with specific release information such as the country, label, barcode and packaging. If you walk into a store and purchase an album or single, they are each represented in MusicBrainz as one release".

If we look at another release-related field in Musicbrainz, we find the "release group" (https://musicbrainz.org/doc/Release_Group):

"A release group, just as the name suggests, is used to group several different releases into a single logical entity. Every release belongs to one, and only one release group.

Both release groups and releases are "albums" in a general sense, but with an important difference: a release is something you can buy as media such as a CD or a vinyl record, while a release group embraces the overall concept of an album -- it doesn't matter how many CDs or editions/versions it had."

By reading these descriptions, we can clearly see that the release group is the table we are looking for as it represents a single creation, no matter how many times it has been edited or released afterwards. So we will have to keep the first release id for each release group.

In [11]:
release_country = pd.read_csv('Musicbrainz/Tables_used/release_country.txt',sep='\t', header=None, engine='c', usecols=[0,1,2])
release_country.columns = ['release_id','area_id','release_year']
release_country.head()

release_id  area_id  release_year
0           3       81        1997.0
1     1427792      107        2014.0
2           9       81        2002.0
3          10      221        2002.0
4          11       81        1999.0

In [12]:
df2 = pd.merge(releases, release_country, how='left', on='release_id')
df2.head()

release_id           release_group  credit_id  area_id  release_year
0           9       A Sorta Fairytale         60     81.0        2002.0
1          10       A Sorta Fairytale         60    221.0        2002.0
2          11       Glory of the 80's         60     81.0        1999.0
3          12  Silent All These Years         60     81.0        1997.0
4          26                  Demons      20211    107.0        1998.0

In [13]:
#Let's see how many releases we have:
df2['release_id'].nunique()

2198457

In [14]:
df2.isnull().sum(axis=0)

release_id            0
release_group         7
credit_id             0
area_id          287376
release_year     341983
dtype: int64

In [15]:
#We want to keep only the releases which have a release year, so we can drop the others:
df2.dropna(subset=['release_year'], axis=0, inplace=True)
df2['release_year'] = df2.release_year.astype(int,inplace=True)
df2['release_id'].nunique()

1859982

In [16]:
#Let's analyze the year column:
pd.options.display.max_rows = 2000
df2.groupby('release_year').count()

release_id  release_group  credit_id  area_id
release_year                                               
1                      2              2          2        2
4                      1              1          1        1
5                      5              5          5        5
7                      1              1          1        1
8                      2              2          2        2
10                     3              3          3        3
14                     1              1          1        1
17                     4              4          4        4
18                     1              1          1        1
19                     3              3          3        3
20                     2              2          2        2
21                     1              1          1        1
23                     1              1          1        1
24                     1              1          1        1
25                     2              2          2        2
28                     2              2          2        2
29                     2              2          2        2
31                     2              2          2        2
73                     1              1          1        1
195                    1              1          1        1
198                    1              1          1        1
199                    2              2          2        2
200                    4              4          4        4
201                    4              4          4        4
209                    2              2          2        2
666                    1              1          1        1
701                    1              1          1        1
996                    1              1          1        1
1009                   1              1          1        1
1029                   1              1          1        1
1198                   1              1          1        1
1694                   1              1          1        1
1886                   1              1          1        1
1889                   1              1          1        1
1890                 249            249        249      249
1891                  11             11         11       11
1892                 130            130        130      130
1893                 183            183        183      183
1894                  29             29         29       29
1895                   2              2          2        2
1896                   6              6          6        6
1897                  65             65         65       65
1898                 183            183        183      183
1899                 281            281        281      281
1900                 131            131        131      131
1901                 199            199        199      199
1902                 234            234        234      234
1903                 196            196        196      196
1904                 155            155        155      155
1905                 163            163        163      163
1906                 154            154        154      154
1907                 171            171        171      171
1908                 212            212        212      212
1909                 187            187        187      187
1910                 115            115        115      115
1911                  54             54         54       54
1912                  37             37         37       37
1913                  13             13         13       13
1914                  22             22         22       22
1915                  12             12         12       12
1916                  26             26         26       26
1917                 135            135        135      135
1918                  87             87         87       87
1919                  46             46         46       46
1920                 102            102        102    

By looking at the different year values, and, in order to have enough values per year, we could drop the rows whose year is below 1890 and above 2019. Our visualization would have 130 years, which is pretty good.

In [17]:
df2.drop(df2[df2['release_year'] < 1890].index , inplace=True)
df2.drop(df2[df2['release_year'] >2019].index , inplace=True)
df2.sort_values(by=['release_year']).head()

release_id                  release_group  credit_id  area_id  \
1266766      386919      Visions of Paradise Waltz      97546    222.0   
1266956      386830  German Ballad with Variations      97546    222.0   
1266958      386829  German Ballad with Variations      97546    222.0   
1266960      386828           Mountain Bells Polka      97546    222.0   
1266961      386827           Mountain Bells Polka      97546    222.0   

         release_year  
1266766          1890  
1266956          1890  
1266958          1890  
1266960          1890  
1266961          1890

In [18]:
#Converting the year column to datetime for later:
df2['release_year'] = pd.to_datetime(df2['release_year'].astype(str), format='%Y')
df2.dtypes

release_id                int64
release_group            object
credit_id                 int64
area_id                 float64
release_year     datetime64[ns]
dtype: object

In [19]:
#We sort by release id and year (we could have 2 release groups with the same name but produced by different artists):
df2.sort_values(['release_group','release_year','credit_id'], ascending=[True,True,True], inplace=True)
df2.head()

release_id            release_group  credit_id  area_id release_year
2026273     2163750                             2205562    240.0   2014-01-01
1648516     1846605                             1503027    240.0   2015-01-01
1250325     1714060    Beaux Soirs De Paris     1324142     73.0   1995-01-01
2116340     2265346                    Le 1     2291833    240.0   2018-01-01
1748061     1895266   M2Music HitDisc Vol. 1          1    222.0   2006-01-01

In [20]:
df2[df2['release_group'] == 'Artaxerxes']

release_id release_group  credit_id  area_id release_year
1836724     2378622    Artaxerxes    2392005    240.0   1996-01-01
1910376     2379252    Artaxerxes    2392005    221.0   2009-01-01
1909444     2379244    Artaxerxes    2392011    222.0   2011-01-01

In [21]:
#Now we can delete the duplicate releases and keep the ones who were first released:
df2.drop_duplicates(subset=['release_group','credit_id'],keep='first', inplace=True)
df2['release_id'].nunique()

1499614

In [22]:
df2[df2['release_group'] == 'Artaxerxes']

release_id release_group  credit_id  area_id release_year
1836724     2378622    Artaxerxes    2392005    240.0   1996-01-01
1909444     2379244    Artaxerxes    2392011    222.0   2011-01-01

## 3) Matching releases with artists

Now that we have both artist and releases dataframes, we can join them:

In [23]:
df3 = pd.merge(df2, df, how='left', on='credit_id')
df3.head()

release_id            release_group  credit_id  area_id release_year  \
0     2163750                             2205562    240.0   2014-01-01   
1     1846605                             1503027    240.0   2015-01-01   
2     1714060    Beaux Soirs De Paris     1324142     73.0   1995-01-01   
3     2265346                    Le 1     2291833    240.0   2018-01-01   
4     1895266   M2Music HitDisc Vol. 1          1    222.0   2006-01-01   

   artist_id                           artist_mbid    artist_name_x  \
0  1654312.0  d10d6441-dcc1-4202-93bf-0c0acf72913a         Soul Glo   
1  1112115.0  7b52c77b-1a34-439d-a285-3a7c69cb5b1a      Ben Bennett   
2  1122795.0  71b8451c-c10a-400e-9544-101f34ab2522  Soixante Étages   
3  1720981.0  a69efb5f-0b28-4328-8ff0-44d8d6f39755         TedeuzeM   
4        1.0  89ad4ac3-39f7-470e-963a-56509c546377  Various Artists   

   start_area1  start_area2  
0       7707.0          NaN  
1          NaN          NaN  
2          NaN          NaN  
3      68613.0          NaN  
4          NaN          NaN

In [24]:
df3.isnull().sum(axis=0)

release_id            0
release_group         4
credit_id             0
area_id               0
release_year          0
artist_id           151
artist_mbid         151
artist_name_x       155
start_area1      430452
start_area2      959581
dtype: int64

In [25]:
df3['release_id'].nunique()

1499614

In [26]:
len(df3)

1724524

In [27]:
df3[df3['release_group']=='Artaxerxes']

release_id release_group  credit_id  area_id release_year  artist_id  \
119493     2378622    Artaxerxes    2392005    240.0   1996-01-01   391603.0   
119494     2378622    Artaxerxes    2392005    240.0   1996-01-01   124232.0   
119495     2378622    Artaxerxes    2392005    240.0   1996-01-01   688791.0   
119496     2378622    Artaxerxes    2392005    240.0   1996-01-01  1129787.0   
119497     2378622    Artaxerxes    2392005    240.0   1996-01-01  1104538.0   
119498     2378622    Artaxerxes    2392005    240.0   1996-01-01   402986.0   
119499     2378622    Artaxerxes    2392005    240.0   1996-01-01   183632.0   
119500     2378622    Artaxerxes    2392005    240.0   1996-01-01    87510.0   
119501     2378622    Artaxerxes    2392005    240.0   1996-01-01  1078968.0   
119502     2379244    Artaxerxes    2392011    222.0   2011-01-01   854064.0   
119503     2379244    Artaxerxes    2392011    222.0   2011-01-01   124232.0   
119504     2379244    Artaxerxes    2392011    222.0   2011-01-01   989288.0   

                                 artist_mbid              artist_name_x  \
119493  e3062782-ab7b-41bc-8e65-aeea16dc1a89              Ian Partridge   
119494  4e7f1926-8704-4545-a1a1-ded91651c884                Thomas Arne   
119495  f34e9da4-2ee7-4f27-aa34-adc5db791bec         Christopher Robson   
119496  c33f733e-2bf4-402b-9455-1a293601a1cd            Patricia Spence   
119497  5680c729-615b-47e2-969e-27a087c572fb              Philippa Hyde   
119498  70af5d9a-c6e0-4fcf-9cde-4d3d00e0fcb0  The Parley of Instruments   
119499  954d1c83-259f-4a25-8878-10c19bb097af             Catherine Bott   
119500  857588a5-b7aa-4f72-a87b-8f03dca60e30                Roy Goodman   
119501  93da7aaa-250b-46e1-b5ef-0ad78d46dc3f       Richard Edgar‐Wilson   
119502  a87f2b39-84c7-4888-935c-d41943bd7971    Classical Opera Company   
119503  4e7f1926-8704-4545-a1a1-ded91651c884                Thomas Arne   
119504  4c05f6e6-7a40-437e-9f12-a90c1da7b4f8                   Ian Page   

        start_area1  start_area2  
119493        221.0       1178.0  
119494        221.0       1178.0  
119495          NaN          NaN  
119496          NaN          NaN  
119497        221.0          NaN  
119498        221.0       1178.0  
119499        221.0          NaN  
119500        221.0      30926.0  
119501          NaN          NaN  
119502        221.0          NaN  
119503        221.0       1178.0  
119504        221.0          NaN

If we look at the above, we can see that there is one line per each artist that participated for each release ID.

As we don't want to show duplicate releases, we need to keep only one artist per release. We will keep the first artist appearing for each release (even though we know this is not 100% accurate, but we have to avoid duplicates). This will afftect 224.910 rows under a total of 1.499.614 unique releases, so 14% of our dataset.

In [28]:
#Now we can delete the duplicate releases and keep the ones who were first released:
df3.drop_duplicates(subset=['release_id'],keep='first', inplace=True)
df3['release_id'].nunique()

1499614

In [29]:
len(df3)

1499614

## 4) Geographical data

The idea of the visualization is to see where each gender comes from, so, ideally, we would have to look at the artists origins (start area: last 2 columns of our dataframe).

In our dataframe df3, the 5th column "area_id" is related to the area where the release was produced. This isn't directly related to the origin of an artist/band, as many artists have to record their works in different countries/or areas.

Let's see for how many releases we have that information:

In [30]:
df3.isnull().sum(axis=0)

release_id            0
release_group         4
credit_id             0
area_id               0
release_year          0
artist_id           151
artist_mbid         151
artist_name_x       155
start_area1      404503
start_area2      876562
dtype: int64

In Musicbrainz's database, we have some tables related to the areas. Let's see how we can use them to input more geographical information into our dataframe:

In [31]:
areas = pd.read_csv('Musicbrainz/Tables_used/area.txt',sep='\t', header=None, engine='python', usecols=[0,2,3])
areas.columns = ['area_id','area_name','code_type']
areas.head()

area_id area_name  code_type
0    15449   Greccio        4.0
1       38    Canada        1.0
2       43     Chile        1.0
3       44     China        1.0
4       36  Cambodia        1.0

In [172]:
#Let's see the area types we have:
area_types = pd.read_csv('Musicbrainz/Tables_used/area_type.txt',sep='\t', header=None, engine='python', usecols=[1,3,4], error_bad_lines=False)
area_types.columns = ['type','code_type','definition']
area_types.head(20)

type  code_type                                         definition
0       Country          1  Country is used for areas included (or previou...
1   Subdivision          2  Subdivision is used for the main administrativ...
2        County          7  County is used for smaller administrative divi...
3  Municipality          4  Municipality is used for small administrative ...
4          City          3  City is used for settlements of any size, incl...
5      District          5  District is used for a division of a large cit...
6        Island          6  Island is used for islands and atolls which do...

In [177]:
#Add the area name and type to our main dataframe for the column "area_id":
df4 = pd.merge(df3, areas, how='left', on='area_id')
df4.head()

release_id                          release_mbid            release_group  \
0     2163750  ca878831-17c3-4a40-88dc-29df8d720248                            
1     1846605  bf301373-4d19-4d51-b44d-23a9c63e9af3                            
2     1714060  8aa72727-4458-49b1-8e4d-f542ac94238f    Beaux Soirs De Paris    
3     2265346  45d5ea0f-c937-41eb-9516-0c059122ed4b                    Le 1    
4     1895266  3849e1c3-1a80-4ad1-a8b0-b889c4f41e67   M2Music HitDisc Vol. 1   

   credit_id  area_id release_year  artist_id  \
0    2205562    240.0   2014-01-01  1654312.0   
1    1503027    240.0   2015-01-01  1112115.0   
2    1324142     73.0   1995-01-01  1122795.0   
3    2291833    240.0   2018-01-01  1720981.0   
4          1    222.0   2006-01-01        1.0   

                            artist_mbid    artist_name_x  start_area1  \
0  d10d6441-dcc1-4202-93bf-0c0acf72913a         Soul Glo       7707.0   
1  7b52c77b-1a34-439d-a285-3a7c69cb5b1a      Ben Bennett          NaN   
2  71b8451c-c10a-400e-9544-101f34ab2522  Soixante Étages          NaN   
3  a69efb5f-0b28-4328-8ff0-44d8d6f39755         TedeuzeM      68613.0   
4  89ad4ac3-39f7-470e-963a-56509c546377  Various Artists          NaN   

   start_area2      area_name  code_type  
0          NaN    [Worldwide]        NaN  
1          NaN    [Worldwide]        NaN  
2          NaN         France        1.0  
3          NaN    [Worldwide]        NaN  
4          NaN  United States        1.0

In [179]:
#Rearranging dataframe columns to have a clearer dataframe:
cols = list(df4.columns.values)
cols

['release_id',
 'release_mbid',
 'release_group',
 'credit_id',
 'area_id',
 'release_year',
 'artist_id',
 'artist_mbid',
 'artist_name_x',
 'start_area1',
 'start_area2',
 'area_name',
 'code_type']